### ISSUE TITLE
Week Three lesson development

### ISSUE THREAD

**Issue by drjwbaker** [LibraryCarpentry/week-three-library-carpentry#1]
Pinging @davanstrien. This will be where we host and work on the week three (Git) lesson.

**Comment by davanstrien** [comment]
Great. Will try and make a start on this soon.

**Comment by drjwbaker** [comment]
Great!

**Comment by drjwbaker** [comment]
@davanstrien: any progress on this? Might be useful to incorporate resources such as https://try.github.io/ and http://pcottle.github.io/learnGitBranching/ into reference materials.

**Comment by davanstrien** [comment]
A little. Have been busier then expected but I promise I will have something up very soon!

**Comment by drjwbaker** [comment]
No worries. In your own time.

**Comment by drjwbaker** [comment]
@jezcope confirmed to assist on 23 November. @davanstrien: any updates?

**Comment by davanstrien** [comment]
Sorry! Buried under work/dissertation. I have scheduled time to work on this tonight and at the weekend so I should have a final draft ready by Sunday evening at the latest.

**Comment by drjwbaker** [comment]
@davanstrien: okay, perfectly understandable. Can we say you'll post Sunday regardless of how far you get, and then we can iterate from there? As ever, huge appreciation for your efforts on this.

**Comment by davanstrien** [comment]
Sounds good, will let you know where I've got to then.

**Comment by jezcope** [comment]
Cool, let me know if there's anything I can do.

**Comment by davanstrien** [comment]
I have made some updates. Hopefully hour one just needs some tidying and it is ready. I made some suggestions for hour two but haven't had time to do handouts / text yet. I can try and get some more done this week but won't have much time. My suggestion would be to focus on push/pulling between two people to start with. Pull requests next and potentially managing conflicts though I think getting to grips with the github pages adding some markdown to the site will probably keep people busy enough

**Comment by drjwbaker** [comment]
Great job @davanstrien: I'll aim to take a proper look this week!

Small thing, setup: we'll need some generic advise for both wk3 and wk2 https://github.com/LibraryCarpentry/week-two-library-carpentry re setup for Windows and OS X users. Windows is covered as for week two I get them to install https://git-for-windows.github.io/ (shell with Git built in!) I don't have a Mac though, so can someone advise on options for OS X please?

**Comment by davanstrien** [comment]
I can produce some instructions for mac. It was a while since I did it but
I seem to remember it being fairly quick and painless.

**Comment by drjwbaker** [comment]
Cheers Daniel. Stick in somewhere in the week three repo and I'll copy it across to the week two repo for completeness.

**Comment by jezcope** [comment]
Software Carpentry have [this page](http://software-carpentry.org/workshops/setup.html) that you should be able to crib from.

**Comment by drjwbaker** [comment]
Great idea @jezcope. Do make sure they are cited appropriately http://software-carpentry.org/license.html (not least because Greg from SC is helping out along the way!)

**Comment by drjwbaker** [comment]
@davanstrien: this is great, super job. Some comments to consider:
- PhD Comics link is broken
- note that I don't cover `touch` week-two so be prepared to explain it
- `open git_test.txt`: note that `open` may not work for everyone depending on their setup... It certainly doesn't work for me on Ubuntu (I tend to do `kate git_text.txt` knowing that `kate` is my text editor, so consider your options here - perhaps just get them to open the text manually, not doing everything no the CL?)
- 'The -u tells Git to remember the parameters'. What 'parameters' do you mean here, be explicit.

On hour two and GitHub Pages, you'd need to produce a clear guide here so that folks can resolve conflicts without getting too frustrated. If this seems too much work, you could split this hour between activities that embed knowledge (eg getting them in groups to draw process diagrams of how Git works?) and you demoing the ease with which a GitHub Pages page could be setup and contributed to (with, say, folks committing to this page from one workstation in groups rather than all the attendees working on collaborative git stuff at the same time! - would certainly help with problem solving in the room).

**Comment by drjwbaker** [comment]
Daniel: @raquel-ucl https://github.com/UCL/rsd-engineeringcourse/tree/master/ch02git has kindly forwarded the Git lesson that @jamespjh uses at UCL. Might be worth taking a look for bits that are of interest.

**Comment by davanstrien** [comment]
thanks both. Should have some more done by the end of this weekend.

**Comment by raquelalegre** [comment]
@davanstrien No problem! Let me know if you need any clarification on the notes.

**Comment by davanstrien** [comment]
Apologies again for the delay - as usual finding it hard to find time for this. I will aim to work on this during the week and have a more final draft done by Friday. 

@drjwbaker Would it be possible to quickly discuss how you want to carry out the actual session during week 1?

**Comment by drjwbaker** [comment]
@davanstrien To be clear what you mean, you'd like to discuss with me content for week three sometime next week?

**Comment by davanstrien** [comment]
Yes, that would be great. I will hopefully have the content ready to go by week 1 but would be good to discuss format etc.

**Comment by drjwbaker** [comment]
@davanstrien: do you have any free time next Monday (9 Nov) afternoon? I'll be in London before Library Carpentry kicks off at 1730, so happy to talk through the week 3 lesson then (just not too close to 1730!)

**Comment by davanstrien** [comment]
Monday afternoon is looking like it might be tricky. I think after the
session I'll have a good sense of how they are running so probably won't
have too many questions.

**Comment by drjwbaker** [comment]
You sure? We can meet in the morning if you want? I won't have much time after as I'll need to shoot off home.

**Comment by jezcope** [comment]
I've been having a look at your `working-branch` branch. This is looking pretty good so far. Here are a handful of comments, mostly minor suggestions. Feel free to disagree with any or all of them!
- I would probably leave introducing github as late as possible (i.e. make a few more commits and look at history and diff first) — introducing `git remote` and `git add` disrupts a nice flow that you've got going by this point, and learners who are new to git are likely to assume you are introducing things in a logical order for a reason and thus that you have to add a remote and push to it before you can do `git log` or `git diff`
- I like the way you've used editing locally and direct on github as a way to simulate collaboration before doing it for real.
- Hours 2 & 3 are shaping up nicely, and setting up a github page looks like a good consolidation project, but bear in mind [James's comments above](https://github.com/LibraryCarpentry/week-three-library-carpentry/issues/1?_pjax=%23js-repo-pjax-container#issuecomment-149829611) and be prepared to be flexible and spend more time than you expect answering questions and clearing up confusion.
- Is it helpful to refer to 'document' so much in the introduction? People will bring their own definition of "document", which may cause confusion. Some people might limit to "something I intend to print" and exclude anything else, while others might include things that don't play well with version control, such as Word and PDF documents. Could maybe use "file"?
- Capitalisation: remember to use lowercase when you are talking about the command (otherwise attendees will be tempted to actually type `Git` and get an error)
- Use backticks (`like this`) to make code (e.g. commands and filenames) stand out inline in markdown.
- Things to consider including:
  - Reverting — i.e. a way to undo something you didn't mean to commit
  - Handling conflicts (this _will_ happen in any collaborative project, so worth making sure people are ready for it when it does)
  - `.gitignore`
  - Demo a GUI — git usually works better when people understand the commandline interface, and there are no GUIs that work reliably enough across platforms for teaching, but it's nice for people to know that there are options.
  - More exploration of history, including `git checkout`
  - Branching

I hope that helps! If yuo have any particular thoughts on how you'd like me to work as a helper, let me know.

**Comment by davanstrien** [comment]
Yeah, think getting time off work at the moment will be tricky. I should
have a better sense of how things work after this evening.

**Comment by davanstrien** [comment]
thanks for the suggestions/corrections. I'll hopefully have time to
incorporate some of them in the next couple of days.

Are you going to be attending the sessions tonight? Maybe we could have a
quick chat about running the session then?

**Comment by jezcope** [comment]
No, I'm based in Leeds/Sheffield so won't be able to make it down to London until week 3 itself. I can do online chat or video though (irc/Hangouts/Skype) if that works for you?

**Comment by davanstrien** [comment]
That would be great @jezcope. Quick Skype chat would probably be useful. Don't think we'd need long but might be quicker than email. Sorry to do doodle poll but maybe easier http://doodle.com/created?pollId=hzpdk82f49ycw2hi&adminKey=ptta96vh If none of those dates work I'm happy to outline my suggestions in an email and go from there.

**Comment by jezcope** [comment]
Sorry, no excuses — I should have got to this sooner. I could probably fit in a chat this evening or tomorrow evening, lunchtime today or Friday morning. Do any of those suit you?

**Comment by drjwbaker** [comment]
Both - can I add that if you need anything printed by last opportunity to do this is Friday, so I'll need it no later than tomorrow.

**Comment by davanstrien** [comment]
My turn to apologise. Last week was very hectic... I will be aiming to get to City about an hour before the session. If you were around we could have a quick chat about the session then? I think the main thing we will have to keep an eye on is people running into difficulties and potentially helping people trouble shoot.

**Comment by jezcope** [comment]
No worries — it's definitely that time of year! My train gets into Kings Cross at half 4 and I'll aim to be at City ASAP after that, probably about 5. Sound ok?

**Comment by raquelalegre** [comment]
Hey there. I'll try and be there around 4:30 and I'm happy to discuss preparation. We had software carpentry at UCL ten days ago and it's still fresh in my memory :)

@davanstrien @jezcope where are you meeting?

**Comment by davanstrien** [comment]
I will probably head straight to the meeting room so I can check projectors etc. Do you know your way around City? I can meet at the entrance at a specified time otherwise.

**Comment by raquelalegre** [comment]
I've never been there before, I'll try to find it otherwise can I phone you? I'm on 07445492624.

**Comment by davanstrien** [comment]
Library-carpentry-git.github.io

**Comment by davanstrien** [comment]
https://github.com/Library-carpentry-git/Library-carpentry-git.github.io

**Comment by raquelalegre** [comment]
This is the command to reload the screen and show the contents of the folder and demystify git's magic:

``` bash
while :
do 
 clear
 tree -a
 sleep 1
done
```

**Comment by davanstrien** [comment]
@drjwbaker @raquel-ucl @jezcope Thanks for all your help yesterday. Was much appreciated.

**Comment by ptallentire** [comment]
Also thanks @davanstrien - it was a really good session.
